src/targets/gpu/device/where.cpp

```diff
@@ -0,0 +1,28 @@
+#include <migraphx/gpu/device/where.hpp>
+#include <migraphx/gpu/device/tensor.hpp>
+#include <migraphx/gpu/device/types.hpp>
+#include <migraphx/gpu/device/launch.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+namespace gpu {
+namespace device {
+
+void where(hipStream_t stream,
+           const argument& result,
+           const argument& arg0,
+           const argument& arg1,
+           const argument& arg2)
+{
+    hip_visit_all(result, arg1, arg2)([&](auto output, auto x, auto y) {
+        hip_visit_all(arg0)([&](auto cond) {
+            gs_launch(stream, arg1.get_shape().elements())(
+                [=](auto i) __device__ { output[i] = cond[i] ? x[i] : y[i]; });
```

Comment:
Actually, this should compute the multi_idx and then pass it to the tensors:

```cpp
auto i = output.get_shape().multi(i);
output[i] = cond[i] ? x[i] : y[i];
```

Reply:
the multi_index i here cannot be used by the cond, we may need another way to access the cond.

Reply:
Why cant the multi_index be used by the condition?

Reply:
I'm not sure myself, but I confirmed that it won't compile that way.

Reply:
@pfultz2 Because the cond and other arguments are from different visit() functions (their types are different), then the indices are considered as different types.

Reply:
I see. The index class is a different type because the tensor ranks could be different. We should skip generating a kernel when tensor ranks are not the same. Perhaps, something like this would work:

```cpp
template<class Shape>
constexpr auto get_rank(const Shape&)
{
    return decltype(typename Shape::hip_index{}.size()){};
}

hip_visit_all(result, arg1, arg2)([&](auto output, auto x, auto y) {
    hip_visit_all(arg0)([&](auto cond) {
        if constexpr(get_rank(cond.get_shape()) == get_rank(output.get_shape()))
        {
            gs_launch(stream, arg1.get_shape().elements())(
                [=](auto i) __device__ { 
                    auto i = output.get_shape().multi(idx);
                    output[i] = cond[i] ? x[i] : y[i];
                });
        }
    });
});
```

 I haven't compiled it yet. We could also consider adding `rank` to the shape class.

Reply:
This works and has been added